File: civil3d/plugin/src/Licensing.cs
// Offline licensing gate — the C# mirror of drawingchecker/licensing.py.
// Both tools read the same %APPDATA%\ARX-Tools\license.lic (RSA-2048
// PKCS#1 v1.5 / SHA-256 signed payload) and derive the same machine
// fingerprint from the Windows MachineGuid, so one license covers the
// Revit extension and this plugin on a machine.
//
// While PublicKeyModulusHex is empty, licensing is OFF (state
// "unconfigured", tools run free). To enforce it, run
// `python3 tools/license_admin.py keygen` and paste the printed constant
// here and in licensing.py, then rebuild.
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Nodes;

namespace ArxChecker.Core;

public class LicenseStatus
{
    public string State;        // unconfigured | valid | grace | missing |
                                // invalid | wrong_machine | expired | clock
    public bool Allowed;
    public bool Warning;
    public string Message;
    public JsonObject Payload;
    public int? DaysLeft;
    public string MachineId;
    public string LicensePath;

    public string Describe()
    {
        var lines = new List<string> { Message };
        var licensee = RulesStore.Str(Payload, "licensee");
        var licenseId = RulesStore.Str(Payload, "licenseId");
        if (!string.IsNullOrEmpty(licensee)) lines.Add("Licensed to: " + licensee);
        if (!string.IsNullOrEmpty(licenseId)) lines.Add("License ID: " + licenseId);
        if (!Allowed)
        {
            if (State == "missing" && LicensePath != null)
            {
                lines.Add("");
                lines.Add("Place your license file at:");
                lines.Add("  " + LicensePath);
            }
            lines.Add("");
            lines.Add("To purchase or fix a license, send this machine ID to support:");
            lines.Add("  Machine ID: " + (MachineId ?? Licensing.MachineFingerprint()));
        }
        return string.Join("\n", lines);
    }
}

public static class Licensing
{
    // Vendor public key — paste the value printed by
    //   python3 tools/license_admin.py keygen
    // Empty = licensing disabled (development / internal builds).
    public const string PublicKeyModulusHex = "";
    public const int PublicKeyExponent = 65537;

    public const string ProductId = "arx-drawing-checker";
    public const string LicenseFormat = "arx-license/1";
    public const int ExpiryWarningDays = 14;
    public const int DefaultGraceDays = 7;
    public const long ClockToleranceSeconds = 2 * 86400;

    // ---- machine fingerprint (identical derivation to the Python side) ----

    public static string FingerprintFromString(string raw)
    {
        var digest = SHA256.HashData(Encoding.UTF8.GetBytes(raw.Trim().ToLowerInvariant()));
        var compact = Convert.ToHexString(digest)[..16];
        return string.Join("-", Enumerable.Range(0, 4).Select(i => compact.Substring(i * 4, 4)));
    }

    public static string MachineFingerprint()
    {
        var overrideId = Environment.GetEnvironmentVariable("ARX_MACHINE_ID");
        if (!string.IsNullOrEmpty(overrideId)) return FingerprintFromString(overrideId);
        if (OperatingSystem.IsWindows())
        {
            try
            {
                var guid = Microsoft.Win32.Registry.GetValue(
                    @"HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography",
                    "MachineGuid", null) as string;
                if (!string.IsNullOrEmpty(guid)) return FingerprintFromString(guid);
            }
            catch { }
        }
        return FingerprintFromString(Environment.MachineName);
    }

    // ---- signature verification ----

    public static bool VerifySignature(byte[] payload, byte[] signature,
                                       string modulusHex, int exponent)
    {
        try
        {
            using var rsa = RSA.Create();
            rsa.ImportParameters(new RSAParameters
            {
                Modulus = Convert.FromHexString(
                    modulusHex.Length % 2 == 1 ? "0" + modulusHex : modulusHex),
                Exponent = ExponentBytes(exponent),
            });
            return rsa.VerifyData(payload, signature, HashAlgorithmName.SHA256,
                                  RSASignaturePadding.Pkcs1);
        }
        catch
        {
            return false;
        }
    }

    private static byte[] ExponentBytes(int exponent)
    {
        var bytes = BitConverter.GetBytes(exponent);
        if (BitConverter.IsLittleEndian) Array.Reverse(bytes);
        var start = Array.FindIndex(bytes, b => b != 0);
        return bytes[(start < 0 ? bytes.Length - 1 : start)..];
    }

    // ---- clock-rollback guard (shares license_state.json with Python) ----

    private static bool CheckClock(string statePath, long now)
    {
        if (statePath == null) return true;
        long lastSeen = 0;
        try
        {
            var state = JsonNode.Parse(File.ReadAllText(statePath, Encoding.UTF8));
            lastSeen = (long)(state?["lastSeen"]?.GetValue<double>() ?? 0);
        }
        catch { }
        if (now + ClockToleranceSeconds < lastSeen) return false;
        if (now > lastSeen)
        {
            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(statePath));
                File.WriteAllText(statePath,
                    new JsonObject { ["lastSeen"] = now }.ToJsonString(),
                    new UTF8Encoding(false));
            }
            catch { }
        }
        return true;
    }

    // ---- license check ----

    private static long? ParseDate(string value)
    {
        if (string.IsNullOrEmpty(value)) return null;
        var date = DateTime.SpecifyKind(
            DateTime.ParseExact(value, "yyyy-MM-dd", null), DateTimeKind.Utc);
        return new DateTimeOffset(date).ToUnixTimeSeconds();
    }

    public static string DefaultLicensePath() =>
        Environment.GetEnvironmentVariable("ARX_LICENSE_FILE")
        ?? AppDataFile("license.lic");

    private static string AppDataFile(string name) =>
        Path.Combine(Environment.GetFolderPath(
            Environment.SpecialFolder.ApplicationData), "ARX-Tools", name);

    private static LicenseStatus Status(string state, string message, bool allowed,
        bool warning = false, JsonObject payload = null, int? daysLeft = null,
        string machineId = null, string licensePath = null) => new()
    {
        State = state, Message = message, Allowed = allowed, Warning = warning,
        Payload = payload, DaysLeft = daysLeft, MachineId = machineId,
        LicensePath = licensePath,
    };

    public static LicenseStatus CheckLicense(string licensePath = null,
        long? nowEpoch = null, string machineId = null,
        string modulusHex = null, string statePath = "<default>")
    {
        modulusHex ??= PublicKeyModulusHex;
        if (string.IsNullOrEmpty(modulusHex))
            return Status("unconfigured",
                "Licensing is not configured; running in unrestricted mode.", true);

        licensePath ??= DefaultLicensePath();
        if (statePath == "<default>") statePath = AppDataFile("license_state.json");
        machineId ??= MachineFingerprint();
        var now = nowEpoch ?? DateTimeOffset.UtcNow.ToUnixTimeSeconds();

        if (licensePath == null || !File.Exists(licensePath))
            return Status("missing", "No license file was found.", false,
                          machineId: machineId, licensePath: licensePath);

        string payloadText;
        byte[] signature;
        try
        {
            var envelope = (JsonObject)JsonNode.Parse(
                File.ReadAllText(licensePath, Encoding.UTF8));
            if (RulesStore.Str(envelope, "format") != LicenseFormat)
                throw new FormatException("unsupported format");
            payloadText = RulesStore.Str(envelope, "payload");
            signature = Convert.FromBase64String(RulesStore.Str(envelope, "signature"));
        }
        catch
        {
            return Status("invalid", "The license file could not be read.", false,
                          machineId: machineId, licensePath: licensePath);
        }

        if (!VerifySignature(Encoding.UTF8.GetBytes(payloadText), signature,
                             modulusHex, PublicKeyExponent))
            return Status("invalid", "The license signature is not valid.", false,
                          machineId: machineId, licensePath: licensePath);

        JsonObject payload;
        try { payload = (JsonObject)JsonNode.Parse(payloadText); }
        catch
        {
            return Status("invalid", "The license payload could not be read.", false,
                          machineId: machineId, licensePath: licensePath);
        }

        if (RulesStore.Str(payload, "product") != ProductId)
            return Status("invalid", "The license is for a different product.", false,
                          payload: payload, machineId: machineId, licensePath: licensePath);

        var machineIds = (payload["machineIds"] as JsonArray)?
            .Select(node => node?.ToString()).Where(s => s != null).ToList() ?? new();
        if (machineIds.Count > 0 && !machineIds.Contains(machineId))
            return Status("wrong_machine", "The license is locked to a different machine.",
                          false, payload: payload, machineId: machineId,
                          licensePath: licensePath);

        if (!CheckClock(statePath, now))
            return Status("clock", "The system clock appears to have been set back; " +
                          "fix the date and time to continue.", false, payload: payload,
                          machineId: machineId, licensePath: licensePath);

        var expires = ParseDate(RulesStore.Str(payload, "expires"));
        if (expires == null)
            return Status("valid", "License valid (perpetual).", true, payload: payload,
                          machineId: machineId, licensePath: licensePath);

        var expiryEnd = expires.Value + 86400; // valid through the whole expiry day
        var graceDays = (int?)(payload["graceDays"] as JsonValue)?.GetValue<double>()
                        ?? DefaultGraceDays;
        var expiresText = RulesStore.Str(payload, "expires");
        if (now < expiryEnd)
        {
            var daysLeft = (int)((expiryEnd - now) / 86400);
            return Status("valid",
                $"License valid — expires in {daysLeft} day(s) on {expiresText}.",
                true, warning: daysLeft <= ExpiryWarningDays, payload: payload,
                daysLeft: daysLeft, machineId: machineId, licensePath: licensePath);
        }
        if (now < expiryEnd + (long)graceDays * 86400)
        {
            var remaining = (int)((expiryEnd + (long)graceDays * 86400 - now) / 86400);
            return Status("grace",
                $"License expired on {expiresText} — running in the renewal grace " +
                $"period ({remaining} day(s) left). Please renew.",
                true, warning: true, payload: payload, daysLeft: remaining,
                machineId: machineId, licensePath: licensePath);
        }
        return Status("expired", $"The license expired on {expiresText}.", false,
                      payload: payload, machineId: machineId, licensePath: licensePath);
    }
}

File: revit/DrawingChecker.extension/lib/drawingchecker/licensing.py
# -*- coding: utf-8 -*-
"""Offline licensing gate shared by the ARX checkers.

A license is a small JSON file (`%APPDATA%\\ARX-Tools\\license.lic`, shared
by the Revit and Civil 3D tools) holding a payload string and an RSA
PKCS#1 v1.5 / SHA-256 signature over that exact string. Only the PUBLIC
key is embedded here; licenses are issued with `tools/license_admin.py`,
which keeps the private key on the vendor's machine. Verification is pure
stdlib (hashlib + modular exponentiation) so it runs identically under
IronPython 2.7 inside Revit and CPython 3.

The payload supports node-locking (machineIds = fingerprints derived from
the Windows MachineGuid), an expiry date with a renewal grace window, and
a clock-rollback guard via a last-seen timestamp in license_state.json.

While PUBLIC_KEY_MODULUS_HEX below is empty, licensing is OFF and every
check returns state 'unconfigured' (allowed) — development/internal
builds keep working with no license at all. To enforce licensing, run
`python3 tools/license_admin.py keygen` and paste the printed constant.
"""
from __future__ import unicode_literals

import base64
import binascii
import hashlib
import io
import json
import os
import time

# ---------------------------------------------------------------------------
# Vendor public key — paste the values printed by
#   python3 tools/license_admin.py keygen
# Empty modulus = licensing disabled (state 'unconfigured', tools run free).
# ---------------------------------------------------------------------------
PUBLIC_KEY_MODULUS_HEX = ''
PUBLIC_KEY_EXPONENT = 65537

PRODUCT_ID = 'arx-drawing-checker'
LICENSE_FORMAT = 'arx-license/1'
LICENSE_FILE_NAME = 'license.lic'
STATE_FILE_NAME = 'license_state.json'
APPDATA_DIR_NAME = 'ARX-Tools'

EXPIRY_WARNING_DAYS = 14      # start warning this many days before expiry
DEFAULT_GRACE_DAYS = 7        # still runs (with a warning) after expiry
CLOCK_TOLERANCE_SECONDS = 2 * 86400  # allowance for timezone/DST changes

# DigestInfo prefix for SHA-256 (EMSA-PKCS1-v1_5, RFC 8017)
_SHA256_PREFIX = binascii.unhexlify(b'3031300d060960864801650304020105000420')

BLOCKED_STATES = ('missing', 'invalid', 'wrong_machine', 'expired', 'clock')


# ---------------------------------------------------------------------------
# RSA PKCS#1 v1.5 / SHA-256 signature verification (pure stdlib)
# ---------------------------------------------------------------------------

def _bytes_to_int(data):
    return int(binascii.hexlify(data), 16)


def _int_to_bytes(value, length):
    hexstr = '%x' % value
    hexstr = hexstr.rjust(length * 2, '0')
    return binascii.unhexlify(hexstr.encode('ascii'))


def verify_signature(payload_bytes, signature_bytes, modulus, exponent):
    """True if signature_bytes is a valid RSASSA-PKCS1-v1_5/SHA-256
    signature of payload_bytes under the given public key."""
    key_len = (modulus.bit_length() + 7) // 8
    if len(signature_bytes) != key_len:
        return False
    decrypted = pow(_bytes_to_int(signature_bytes), exponent, modulus)
    encoded = _int_to_bytes(decrypted, key_len)
    digest = hashlib.sha256(payload_bytes).digest()
    padding_len = key_len - len(_SHA256_PREFIX) - len(digest) - 3
    if padding_len < 8:
        return False
    expected = (b'\x00\x01' + b'\xff' * padding_len + b'\x00'
                + _SHA256_PREFIX + digest)
    return encoded == expected


# ---------------------------------------------------------------------------
# Machine fingerprint (node-lock)
# ---------------------------------------------------------------------------

def fingerprint_from_string(raw):
    """XXXX-XXXX-XXXX-XXXX from any machine-unique string. The C# plugin
    implements the same derivation, so one license covers both tools."""
    digest = hashlib.sha256(raw.strip().lower().encode('utf-8')).hexdigest()
    compact = digest[:16].upper()
    return '-'.join(compact[i:i + 4] for i in range(0, 16, 4))


def _windows_machine_guid():
    try:
        try:
            import winreg  # CPython 3 on Windows
        except ImportError:
            import _winreg as winreg  # IronPython 2.7
        key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE,
                             r'SOFTWARE\Microsoft\Cryptography')
        try:
            value, _kind = winreg.QueryValueEx(key, 'MachineGuid')
            return value
        finally:
            winreg.CloseKey(key)
    except Exception:
        return None


def machine_fingerprint():
    """Fingerprint of this machine. ARX_MACHINE_ID overrides (tests,
    virtual desktops); otherwise the Windows MachineGuid; last-resort
    fallback is the hostname + MAC so non-Windows dev boxes get a value."""
    override = os.environ.get('ARX_MACHINE_ID')
    if override:
        return fingerprint_from_string(override)
    guid = _windows_machine_guid()
    if guid:
        return fingerprint_from_string(guid)
    import platform
    import uuid
    return fingerprint_from_string('%s|%s' % (platform.node(), uuid.getnode()))


# ---------------------------------------------------------------------------
# Locations
# ---------------------------------------------------------------------------

def _appdata_path(file_name):
    appdata = os.environ.get('APPDATA')
    if not appdata:
        return None
    return os.path.join(appdata, APPDATA_DIR_NAME, file_name)


def default_license_path():
    return os.environ.get('ARX_LICENSE_FILE') or _appdata_path(LICENSE_FILE_NAME)


def default_state_path():
    return _appdata_path(STATE_FILE_NAME)


# ---------------------------------------------------------------------------
# Clock-rollback guard
# ---------------------------------------------------------------------------

def _check_clock(state_path, now):
    """Returns False if the clock appears to have been wound back past the
    tolerance; records the newest time seen. Unwritable state is ignored —
    the guard degrades gracefully rather than blocking paying users."""
    if not state_path:
        return True
    last_seen = 0
    try:
        with io.open(state_path, 'r', encoding='utf-8') as fh:
            last_seen = float(json.load(fh).get('lastSeen', 0))
    except Exception:
        pass
    if now + CLOCK_TOLERANCE_SECONDS < last_seen:
        return False
    if now > last_seen:
        try:
            directory = os.path.dirname(state_path)
            if directory and not os.path.isdir(directory):
                os.makedirs(directory)
            with io.open(state_path, 'w', encoding='utf-8') as fh:
                fh.write(json.dumps({'lastSeen': now}))
        except Exception:
            pass
    return True


# ---------------------------------------------------------------------------
# License check
# ---------------------------------------------------------------------------

def _parse_date(value):
    """'YYYY-MM-DD' -> epoch seconds at UTC midnight, or None."""
    if not value:
        return None
    import calendar
    return calendar.timegm(time.strptime(value, '%Y-%m-%d'))


def _status(state, message, allowed, warning=False, payload=None,
            days_left=None, machine_id=None, license_path=None):
    return {
        'state': state,
        'allowed': allowed,
        'warning': warning,
        'message': message,
        'payload': payload,
        'daysLeft': days_left,
        'machineId': machine_id,
        'licensePath': license_path,
    }


def check_license(license_path=None, now=None, machine_id=None,
                  modulus_hex=None, exponent=None, state_path=None):
    """Validate the license and return a status dict:

    state    unconfigured | valid | grace | missing | invalid |
             wrong_machine | expired | clock
    allowed  True when the tool may run (unconfigured / valid / grace)
    warning  True when allowed but the user should be told (grace,
             or valid but expiring within EXPIRY_WARNING_DAYS)

    All parameters default to the embedded key and the shared
    %APPDATA%\\ARX-Tools files; tests inject their own.
    """
    modulus_hex = PUBLIC_KEY_MODULUS_HEX if modulus_hex is None else modulus_hex
    exponent = PUBLIC_KEY_EXPONENT if exponent is None else exponent
    if not modulus_hex:
        return _status('unconfigured', 'Licensing is not configured; running '
                       'in unrestricted mode.', allowed=True)
    modulus = int(modulus_hex, 16)

    if license_path is None:
        license_path = default_license_path()
    if state_path is None:
        state_path = default_state_path()
    if machine_id is None:
        machine_id = machine_fingerprint()
    if now is None:
        now = time.time()

    if not license_path or not os.path.isfile(license_path):
        return _status('missing', 'No license file was found.', allowed=False,
                       machine_id=machine_id, license_path=license_path)

    try:
        with io.open(license_path, 'r', encoding='utf-8-sig') as fh:
            envelope = json.load(fh)
        payload_text = envelope['payload']
        signature = base64.b64decode(envelope['signature'])
        if envelope.get('format') != LICENSE_FORMAT:
            raise ValueError('unsupported format')
    except Exception:
        return _status('invalid', 'The license file could not be read.',
                       allowed=False, machine_id=machine_id,
                       license_path=license_path)

    if not verify_signature(payload_text.encode('utf-8'), signature,
                            modulus, exponent):
        return _status('invalid', 'The license signature is not valid.',
                       allowed=False, machine_id=machine_id,
                       license_path=license_path)

    try:
        payload = json.loads(payload_text)
    except Exception:
        return _status('invalid', 'The license payload could not be read.',
                       allowed=False, machine_id=machine_id,
                       license_path=license_path)

    if payload.get('product') != PRODUCT_ID:
        return _status('invalid', 'The license is for a different product.',
                       allowed=False, payload=payload, machine_id=machine_id,
                       license_path=license_path)

    machine_ids = payload.get('machineIds') or []
    if machine_ids and machine_id not in machine_ids:
        return _status('wrong_machine', 'The license is locked to a '
                       'different machine.', allowed=False, payload=payload,
                       machine_id=machine_id, license_path=license_path)

    if not _check_clock(state_path, now):
        return _status('clock', 'The system clock appears to have been set '
                       'back; fix the date and time to continue.',
                       allowed=False, payload=payload, machine_id=machine_id,
                       license_path=license_path)

    expires = _parse_date(payload.get('expires'))
    if expires is None:
        return _status('valid', 'License valid (perpetual).', allowed=True,
                       payload=payload, machine_id=machine_id,
                       license_path=license_path)

    expiry_end = expires + 86400  # valid through the whole expiry day
    grace_days = payload.get('graceDays')
    if grace_days is None:
        grace_days = DEFAULT_GRACE_DAYS
    if now < expiry_end:
        days_left = int((expiry_end - now) // 86400)
        warn = days_left <= EXPIRY_WARNING_DAYS
        message = ('License valid — expires in %d day(s) on %s.'
                   % (days_left, payload.get('expires')))
        return _status('valid', message, allowed=True, warning=warn,
                       payload=payload, days_left=days_left,
                       machine_id=machine_id, license_path=license_path)
    if now < expiry_end + grace_days * 86400:
        remaining = int((expiry_end + grace_days * 86400 - now) // 86400)
        message = ('License expired on %s — running in the renewal grace '
                   'period (%d day(s) left). Please renew.'
                   % (payload.get('expires'), remaining))
        return _status('grace', message, allowed=True, warning=True,
                       payload=payload, days_left=remaining,
                       machine_id=machine_id, license_path=license_path)
    return _status('expired', 'The license expired on %s.'
                   % payload.get('expires'), allowed=False, payload=payload,
                   machine_id=machine_id, license_path=license_path)


def describe(status):
    """User-facing text for the gate dialog, including everything a
    customer needs to send to support (machine ID, expected path)."""
    lines = [status['message']]
    payload = status.get('payload') or {}
    if payload.get('licensee'):
        lines.append('Licensed to: %s' % payload['licensee'])
    if payload.get('licenseId'):
        lines.append('License ID: %s' % payload['licenseId'])
    if not status['allowed']:
        if status['state'] == 'missing' and status.get('licensePath'):
            lines.append('')
            lines.append('Place your license file at:')
            lines.append('  %s' % status['licensePath'])
        lines.append('')
        lines.append('To purchase or fix a license, send this machine ID '
                      'to support:')
        lines.append('  Machine ID: %s' % (status.get('machineId') or
                                           machine_fingerprint()))
    return '\n'.join(lines)
